File: languages/javascript/advent-of-code/2023/one.js
const fs = require("fs");
const path = require("path");

const readlines = (filename) => {
  // let filepath = path.join(__dirname, filename);
  return fs
    .readFileSync(filename, "utf8", (err, data) => {
      if (err) {
        console.error(err);
        return;
      }
      return data;
    })
    .trim();
};

// pass the input filename as command line argument
let filename = process.argv.slice(2)[0]
if (!filename) {
  throw new Error('Input filename is required with full path')
} 
let filedata = readlines(filename);
let lines = filedata.split("\n");
let sum = 0;

for (let j = 0; j < lines.length; j++) {
  let first_char;
  let last_char;
  let line = lines[j];

  for (let i = 0; i < line.length; i++) {
    if (!isNaN(parseInt(line[i]))) {
      first_char = line[i];
      break;
    }
  }

  for (let k = line.length - 1; k >= 0; k--) {
    if (!isNaN(parseInt(line[k]))) {
      last_char = line[k];
      break;
    }
  }

  let number = parseInt(first_char + last_char);
  sum += number;
}

console.log(sum);

File: languages/rust/advent-of-code/2023/one.rs
// Read file from the system

use std::env;
use std::fs;
use std::path::PathBuf;

fn main() {
    let args: Vec<String> = env::args().collect();
    let file_name = &args[1];

    let current_dir: PathBuf = match env::current_dir() {
        Ok(path) => path,
        Err(e) => {
            println!("Error getting the current directory: {}", e);
            return;
        }
    };

    let file_path = current_dir.join(file_name);
    let content = match fs::read_to_string(&file_path) {
        Ok(content) => content,
        Err(e) => {
            println!("Unable to read the content of file: {}", file_name);
            return;
        }
    };


    let mut sum: i32 = 0;
    for line in content.lines() {
        let mut first_digit: char = '\0';
        let mut second_digit: char = '\0';
        for ch in line.chars() {
            if ch.is_digit(10) {
                first_digit = ch;
                break;
            };
        };

        for ch in line.chars().rev() {
            if ch.is_digit(10) {
                second_digit = ch;
                break;
            };
        };
        match format!("{}{}", first_digit, second_digit).parse::<i32>() {
            Ok(value) => sum += value,
            Err(e) => return,
        }
    };
    println!("{}", sum)
}
